test/test_pipeline_schedule.py

```diff
@@ -0,0 +1,238 @@
+
+    def forward(self, x):
+        a = self.wi(x)
+        a = self.wh1(a)
+        a = self.wh2(a)
+        a = self.wh3(a)
+        b = self.gelu_act(a)
+        c = self.wo(b)
+        return c
+
+
+def setup(local_rank, world_size):
+    # If this is a child process (i.e., its PID is not the same as the PID of the process that started this script)
+    if os.getppid() != os.getpid():
+        set_up_logging(local_rank)
+
+    # initialize the process group
+    logger.info(f"init for rank {local_rank}")
+    dist.init_process_group("nccl")
+    if torch.distributed.is_initialized():
+        torch.cuda.set_device(local_rank)
+
+    logger.info(f"finish init for rank {local_rank}")
+
+
+def main(**kwargs):
+    torch.manual_seed(42)
+    print(f"MY KWARGS ARE {kwargs}")
+    rank = kwargs["rank"]
+    local_rank = kwargs["local_rank"]
+    world_size = kwargs["world_size"]
+
+    setup(local_rank, world_size)
+    logger.info(
+        f"====== World Rank {rank}, Local Rank {local_rank}, World Size {world_size} main ======"
+    )
+
+    input_dim = 4000
+    hidden_dim = 8000
+    output_dim = 4000
+
+    module_list = torch.nn.ModuleList(
+        modules=[
+            MLP(input_dim, hidden_dim, output_dim) for i in range(world_size)
+        ]
+    )
+    microbatch_size = 8
+    global_batch_size = 64
+    assert global_batch_size % microbatch_size == 0
+    n_microbatches = int(global_batch_size / microbatch_size)
+    n_pp = world_size
+
+    x = torch.randn([microbatch_size, input_dim]).to("meta")
+
+    stage_model = PipelineStage(
+        module_list[rank], rank, world_size, rank, world_size, x
+    )
+    stage_model.init_p2p_neighbors()
+
+    stage_model_looped = [
+        PipelineStage(
+            module_list[rank],
+            stage_id=(world_size * i) + rank,
+            num_stages=world_size * world_size,
+            rank=rank,
+            world_size=world_size,
+            meta_input=x,
+        )
+        for i in range(world_size)
+    ]
+    x_cuda_empty = torch.empty_like(x, device="cuda")
+    microbatches = [
+        torch.randn_like(x_cuda_empty) for _ in range(n_microbatches)
+    ]
+
+    for schedule in kwargs["schedules"]:
+        logger.info(f"====== Rank {rank} running schedule {schedule} ======")
+        if schedule == "gpipe":
+            pipeline = PipelineScheduleGPipe(stage_model)
+        elif schedule == "looped_bfs":
+            pipeline = PipelineScheduleLoopedBFS(stage_model_looped)
+        elif schedule == "looped_dfs":
+            pipeline = PipelineScheduleLoopedDFS(
+                stage_model_looped,
+                n_microbatch=n_microbatches,
+                pp_id=rank,
+                n_pp=n_pp,
+            )
+
+        logger.info(f"====== Rank {rank} profile ======")
+
+        # with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]) as prof:
+        #    with record_function(schedule):
+        pipeline.step(microbatches)
```

Comment:
Curious: in reality/more formalized code, would every rank step on `microbatches` or just the first rank? Just to understand the API better.

Reply:
I will clean this up, we expect each rank to call step but the microbatches argument is optional except for the first stage/rank which must pass in the microbatches